### DISCUSSION TITLE
Status of adaptive sampling method

### DISCUSSION THREAD

**Discussion by wvangeit** [snl-dakota/dakota#130] [Q&A]
Hi,
Is anybody actively working on the adaptive sampling method? (https://snl-dakota.github.io/docs/6.19.0/users/usingdakota/reference/method-adaptive_sampling.html)

I know it is marked as experimental, and we wonder if it will become production-ready at some point?
We have trouble at the moment with enabling batch mode together with adaptive sampling.
It does correctly batch-evaluate the initial points, but it looks like that after that it doesn't provide a batch anymore to the evaluation function, but instead a single value, after which it crashes.
(Disabling the batch mode, and adapting the evaluation function accordingly, 'works' without crashing the run)

**Comment by cfd0** [comment]
Dear @wvangeit ,

It's pity nobady answered you yet.
I would like to give it a try to build surrogate models... do you know if there is any tutorial or example available for this method?

Thanks

**Reply by thenectarine** [reply]
1. The behavior described in the original post seems reasonable for an experimental capability. Most new adaptive sampling algorithms start simple by running a batch of initial points and then querying/running the black box function one point at a time during each adaptive iteration. Other algorithms that have had more time for development are capable of querying multiple points during one adaptive iteration.
2. I found this example, available in the Dakota files and linked below, that demonstrates adaptive sampling: dakota_uq_textbook_adaptsampling.in. This example uses predictive variance as the fitness metric. Feel free to reach out if you have more questions on Gaussian process (GP) models. I've worked with GPs to predict deflection, stress, buckling load factors, dynamic and acoustic responses for 1-10 parameters.

[https://github.com/snl-dakota/dakota/blob/devel/test/dakota_uq_textbook_adaptsampling.in](https://github.com/snl-dakota/dakota/blob/devel/test/dakota_uq_textbook_adaptsampling.in)

**Comment by cfd0** [comment]
Hi @thenectarine ,

Thanks a lot for your reply and the link.

I'm quite new to engineering optimization and surrogates and I'm trying to use dakota for obtaining a surrogate of a cfd model with 9 input params and 3  output vars.

I'm not looking for surrogate (point) optimization, but rather to improve the global accuracy of the surrogate with respect to the underlying cfd model, and, to my current understanding, adaptive sampling and GPs looked interesting tools for the purpose...

Up to now, I've run several studies with different sampling methods (random, lhs, sobol) and I've then built a python regression model with those results.  

Looking at the dakota example you suggested, it seems that they are using one response function, for uq purposes... how would you modify it if you wanted to perform a study where, given an initial number of samples (e.g. a prior lhs sampling) you wanted to add new sampling points where the expected variance was the greatest?

**Reply by thenectarine** [reply]
I can confirm that Dakota's adaptive_sampling capability will improve the global accuracy for N response functions. See my investigation below.

If you want to manually supply your own training data, then have Dakota do the adaptive sampling, this is what I have seen in my few days using Dakota.

1. adaptive_sampling is still in development and is documented as an experimental capability. I have seen no example, someone please correct me, where you can provide your own manually acquired training data, then start Dakota at the adaptive sampling portion of the procedure. Dakota always wants to start by initially acquiring its own training data, then perform the adaptive sampling.
2. There is a fancy trick you can try where you supply your own manually acquired training data, then have Dakota do the adaptive sampling. It seems the fancy trick has worked for another individual. I document the fancy trick in this comment: [https://github.com/orgs/snl-dakota/discussions/146#discussioncomment-10975924](https://github.com/orgs/snl-dakota/discussions/146#discussioncomment-10975924) .


**Investigation of Dakota's Adaptive Sampling Capability**

1. I read the Dakota Reference Manual's section on adaptive sampling.
3. I also experimented with the predictive_variance fitness metric and the text book problem. See study_a.in and app.py below.

The procedure started with an initial 20 run LHS. The next 50 runs were dedicated to improving the fitness_metric=predicted_variance of response r1. The next 50 runs were dedicated to improving the fitness_metric=predicted_variance of response r2. A total of 120 runs were performed. The figures below show the predicted function and corresponding variance of response r1. Both response r1 and r2 are the text book function, but r1 depends on variables x1 and x2 and r2 depends on x3 and x4. The plots for r1 and r2 are similar. The goal was to investigate confirm Dakota improves the prediction accuracy of 2 or more response functions and Dakota is confirmed to do this.

Figure 1 - Kriging Model of Response r1
![Screenshot from 2024-10-18 08-00-29](https://github.com/user-attachments/assets/8be8e490-e936-4441-a585-3d33b5b96b50)


Figure 2 - Kriging Model's Variance of Response r1
![Screenshot from 2024-10-18 08-21-04](https://github.com/user-attachments/assets/edc4d3d6-ab23-45b5-a7ad-db644eba51c5)


study_a.in

```
environment
   error_file 'dakota.err'
   output_file 'dakota.out'
   tabular_data
      tabular_data_file 'tabular.data'

method
      adaptive_sampling
         batch_selection
               naive
         fitness_metric
               predicted_variance
         initial_samples 20
         max_iterations 50
         refinement_samples  2
         samples_on_emulator 400
         seed 1234

variables
   uniform_uncertain 4
      descriptors  'x1'  'x2'  'x3'  'x4'
      lower_bounds  -2.0  -2.0  -2.0  -2.0
      upper_bounds  2.0  2.0  2.0  2.0

interface
   analysis_drivers  'python3 app.py'
         fork
            parameters_file 'params.in'
            results_file 'results.out'

responses
      no_gradients
      no_hessians
      response_functions 2
```

app.py
```
import subprocess
import os
import re
import sys
import csv
from io import StringIO


def determine_which_response_labels_to_extract_responses_for(path_of_file):
    # # Use for testing
    # path_of_file = 'params.in'
    # print(determine_which_response_labels_to_extract_responses_for(path_of_file))
    # # returns
    # # ['r1', 'r2', etc.]

    # Comments
    # 1. File params.in has this format
    #                                          1  variables
    #                       2.808069235193960e+02 x1
    #                                          3  functions
    #                                           1 ASV_1:r1
    #                                           0 ASV_2:r2
    #                                           0 ASV_3:r3
    #                                          1  derivative_variables
    #                                           1 DVV_1:x1
    #                                          1  DVV_1:x1
    #                                           0 analysis_components
    #                                           1 eval_id
    #                                           0 metadata
    #    Note the leading 1 in front of ASV_1:r1 means this response should be added to results.out
    #    The other responses have a leading 0 in front of ASV_i:ri, meaning this response
    #    should not be added. This logic determines which labels should be extracted.

    with open(path_of_file) as f:
        lines_of_file = f.readlines()

        for i, item_i in enumerate(lines_of_file):
            lines_of_file[i] = re.sub(r'\s+', ',', item_i.strip())

        lines_of_file.insert(0, 'column_1,column_2')

        incoming_string_of_csv = '\n'.join(lines_of_file)

        a, header = convert_csv_string_to_list_of_objects(incoming_string_of_csv)

        keep_elements = False
        list_of_response_labels = []

        for item_i in a:
            if bool(re.search(r'derivative_variables', item_i['column_2'])):
                keep_elements = False

            if keep_elements is True:
                if int(item_i['column_1']) > 0:
                    list_of_response_labels.append(re.sub(r'ASV_\d+:', '', item_i['column_2']))

            if bool(re.search(r'functions', item_i['column_2'])):
                keep_elements = True

        return list_of_response_labels


def convert_csv_string_to_list_of_objects(incoming_string_of_csv):
    f = StringIO(incoming_string_of_csv)
    reader = csv.reader(f, delimiter=',')
    header = next(reader)

    for iteration, v in enumerate(header):
        header[iteration] = header[iteration].lower()

    a = [dict(zip(header, row)) for row in reader]

    return a, header


def get_parameter_values_from_params_in(path_of_file):
    #     # Use for testing
    #     # This will create a dummy file test_params.in
    #     from tabulate import tabulate
    #     string_params_in = '''                                         10 variables
    #                       7.949322141496712e-02 x1
    #                       8.008899754265753e-02 x2
    #                       8.060040941840664e-02 x3
    #                       8.063159683628703e-02 x4
    #                       7.977830277910684e-02 x5
    #                       8.086283010290178e-02 x6
    #                       7.884785318235264e-02 x7
    #                       7.913206385642148e-02 x8
    #                       7.979288702234376e-02 x9
    #                       7.947816881356040e-02 x10
    #                                          10 functions
    #                                           3 ASV_1:r1
    #                                           3 ASV_2:r2
    #                                           3 ASV_3:r3
    #                                           3 ASV_4:r4
    #                                           3 ASV_5:r5
    #                                           3 ASV_6:r6
    #                                           3 ASV_7:r7
    #                                           3 ASV_8:r8
    #                                           3 ASV_9:r9
    #                                           3 ASV_10:r10
    #                                          10 derivative_variables
    #                                           1 DVV_1:x1
    #                                           2 DVV_2:x2
    #                                           3 DVV_3:x3
    #                                           4 DVV_4:x4
    #                                           5 DVV_5:x5
    #                                           6 DVV_6:x6
    #                                           7 DVV_7:x7
    #                                           8 DVV_8:x8
    #                                           9 DVV_9:x9
    #                                          10 DVV_10:x10
    #                                           0 analysis_components
    #                                          10 eval_id
    #                                           0 metadata
    # '''
    #     with open('test_params.in', 'w') as f:
    #         f.write(string_params_in)
    #     list_z = get_parameter_values_from_params_in('test_params.in')
    #     # Use .15e to display the full number in the table below
    #     print(tabulate(list_z, headers={'column_1': 'label', 'column_2': 'value'}, tablefmt='orgtbl', floatfmt='.15e'))
    #     # returns
    #     # | label   |                 value |
    #     # |---------+-----------------------|
    #     # | x1      | 7.949322141496712e-02 |
    #     # | x2      | 8.008899754265753e-02 |
    #     # | x3      | 8.060040941840664e-02 |
    #     # | x4      | 8.063159683628703e-02 |
    #     # | x5      | 7.977830277910684e-02 |
    #     # | x6      | 8.086283010290178e-02 |
    #     # | x7      | 7.884785318235264e-02 |
    #     # | x8      | 7.913206385642148e-02 |
    #     # | x9      | 7.979288702234376e-02 |
    #     # | x10     | 7.947816881356040e-02 |

    parameter_values = []

    with open(path_of_file) as f:
        lines_of_file = f.readlines()

        for i, item_i in enumerate(lines_of_file):
            lines_of_file[i] = re.sub(r'\s+', ',', item_i.strip())

        lines_of_file.insert(0, 'column_1,column_2')

        incoming_string_of_csv = '\n'.join(lines_of_file)

        a, header = convert_csv_string_to_list_of_objects(incoming_string_of_csv)

        keep_elements = False

        for item_i in a:
            if bool(re.search(r'^functions$', item_i['column_2'])):
                keep_elements = False

            if keep_elements is True:
                parameter_values.append({'label': item_i['column_2'], 'value': item_i['column_1']})

            if bool(re.search(r'^variables$', item_i['column_2'])):
                keep_elements = True

    return parameter_values


if __name__ == '__main__':
    # Get the names of the Dakota input and output files
    name_of_dakota_input_file = sys.argv[1]
    name_of_dakota_output_file = sys.argv[2]

    parameter_values = get_parameter_values_from_params_in(name_of_dakota_input_file)

    response_f_1 = (float(parameter_values[0]['value']) - 1)**4 + (float(parameter_values[1]['value']) - 1)**4
    response_f_2 = (float(parameter_values[2]['value']) - 1)**4 + (float(parameter_values[3]['value']) - 1)**4

    string_to_write_out = ''
    string_to_write_out = string_to_write_out + str(response_f_1) + '\n'
    string_to_write_out = string_to_write_out + str(response_f_2) + '\n'

    with open(name_of_dakota_output_file, 'w', encoding='utf8') as file:
        file.write(string_to_write_out)
```

**Reply by thenectarine** [reply]
Bonus round

Just in case someone else is going down the rabbit hole of adaptive sampling, I do have some opinions.

Opinions

For a 9 parameter problem, adaptive sampling where you try to improve the global accuracy will most likely require hundreds or thousands of runs. This is analogous to putting a street light every 100 meters in the forest until the forest is decently lit at night, i.e. wherever it is dark or there is a high uncertainty, acquire training data for that region. For a 9 parameter problem, you will have so many dark pockets, acquiring training data at all these dark pockets is analogous to putting up enough street lights so the universe is sufficiently lit.


There are some shortcuts to build a good kriging model without having to resort to adaptive sampling and thousands of runs.

Bounds - Kriging models have very poor prediction performance at the bounds because the regions near the bounds are undersampled. LHS generates sample points in the interior of the design space. The trick is to slightly expand your bounds when building your initial LHS so that the new LHS points better cover your previously undersampled regions. Also, be careful with very large bounds. If the bounds are too large such that your response function changes drastically within the bounds, this can throw off kriging models if you don't have enough samples. 

Bayesian Optimization (BO) - If your problem can be formulated into an optimization problem, I recommend trying Bayesian optimization (BO). BO cleverly finds the optimum when the kriging model does not necessarily have good global accuracy, so you wouldn't need to allocate thousands of runs just for the sake of global accuracy.

Leave-one-out test - I don't do a lot of surrogate modeling, I do more BO, but the leave-one-out test is used often when crafting kriging models. The idea with this test is to see if the prediction of the kriging model changes drastically when you leave out one sample point i from the training data. This is repeated for each sample point i in 1 to N, where N is the number of samples.

Convergence Study - I have tried this informally until I decided to rely on BO more than just surrogate modeling alone. Run your black box function for M points and call this data "verification points" and "verification responses." Configure Dakota to perform an incremental sampling procedure for 25, 50, 100, 200, 400, 800, etc. runs. Create kriging models for 25 training data points, 50 points, 100, 200, etc. Each time, take your verification points and evaluate them with the kriging model. Compare the predicted responses with your verification responses and calculate the normalized root mean square error (NRMSE). Repeat for each sampling size. If the NRMSE is converging, this is indication your response function is being well modeled by the kriging model. If the NRMSE is not converging, this could be indication the true response function is horribly rough and non-smooth, outlier sampling points are distorting the kriging model, or there is some other unknown issue.

When I first started exploring surrogate models, I realized they are extremely cumbersome. I try to use BO as much as possible because BO is fully automated. I typically start my BO and run off to lunch. If I have to use only a surrogate model, I skip adaptive sampling and just run the black box function hundreds or thousands of times.

**Reply by cfd0** [reply]
Thank you very much for taking the time to write such an extensive and detailed response and for providing the snippets.
Next days I'll try out the trick you suggested with dummy cheap functions.

Really appreciated also your comments/opinions.

Currently I am doing a convergence study similar to what you suggested, using fixed amount of points and using incremental runs with 32, 64, 128 points for different sampling methods. It will take some time since every "point" is a cfd run. Let's see if I'll be subject to the curse of dimensionality...

**Reply by thenectarine** [reply]
Another approach worth trying is the classical least squares regression with linear or quadratic terms. Least squares regression is also one of the easiest methods available. Below are some examples.

If the true response function(s) is decently smooth and not too wild, you might get lucky with least squares regression.

[ls_linear.py.txt](https://github.com/user-attachments/files/17439281/ls_linear.py.txt)

[ls_quadratic.py.txt](https://github.com/user-attachments/files/17439283/ls_quadratic.py.txt)

**Reply by cfd0** [reply]
Still playing around with adaptive_sampling for surrogate models...  

Maybe I've found a way to reuse previous data points.

It seems to me to be working:  the "import_build_points_file" option should do the trick.
Here below I attach a test case starting from a link you previously suggested.

(Nevertheless for my real case I'm still stuck with dakota asking for of an explicit response function, which for a global surrogate I don't have...)

[dakota_uq_textbook_adaptsampling.in.txt](https://github.com/user-attachments/files/17775503/dakota_uq_textbook_adaptsampling.in.txt)
[output.dat.txt](https://github.com/user-attachments/files/17775504/output.dat.txt)
